Bu kod, XOR problemini çözmek için iki katmanlı (bir gizli katman ve bir çıktı katmanı) bir sinir ağı kullanarak delta öğrenme kuralını uygulamaktadır.

In [ ]:
import numpy as np

# Define the activation function (sigmoid)
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Define the derivative of the activation function
def sigmoid_derivative(x):
    return x * (1 - x)


* sigmoid(x) fonksiyonu, giriş değeri x için sigmoid aktivasyon fonksiyonunu hesaplar.
* sigmoid_derivative(x) fonksiyonu, sigmoid fonksiyonunun türevini hesaplar. Bu türev, geri yayılım sırasında hatanın ağırlıklara nasıl etki ettiğini hesaplamak için kullanılır.

In [ ]:

# Define the delta learning rule function for a two-layered network
def delta_learning_rule_two_layer(input_data, target_output, hidden_neurons, learning_rate, epochs):
    # Initialize random weights
    np.random.seed(1)
    giris_sayisi=2
    synaptic_weights1 = 2 * np.random.random((giris_sayisi, hidden_neurons)) - 1
    synaptic_weights2 = 2 * np.random.random((hidden_neurons, 1)) - 1
    for epoch in range(epochs):
        # Forward propagation
        hidden_layer = sigmoid(np.dot(input_data, synaptic_weights1))
        output_layer = sigmoid(np.dot(hidden_layer, synaptic_weights2))

        # Calculate the error
        error = target_output - output_layer

        # Backpropagation
        output_delta = error * sigmoid_derivative(output_layer)
        hidden_error = output_delta.dot(synaptic_weights2.T)
        hidden_delta = hidden_error * sigmoid_derivative(hidden_layer)

        # Update the weights
        synaptic_weights2 += learning_rate * hidden_layer.T.dot(output_delta)
        synaptic_weights1 += learning_rate * input_data.T.dot(hidden_delta)

    return synaptic_weights1, synaptic_weights2




   - Girdi verilerini, hedef çıktıları, gizli katman nöron sayısını, öğrenme oranını ve epoch (iterasyon) sayısını parametre olarak alır.
   - Rastgele ağırlıkları başlatır.
   - Belirtilen epoch sayısı kadar eğitim döngüsü gerçekleştirir.
   - İleri yayılım yaparak gizli katman ve çıktı katmanı çıktıları hesaplanır.
   - Hata hesaplanır ve geri yayılım ile ağırlıklar güncellenir.

In [ ]:
# Define the XOR problem data
input_data = np.array([[0, 0],
                       [0, 1],
                       [1, 0],
                       [1, 1]])

target_output = np.array([[0, 1, 1, 0]]).T

# Set the number of hidden neurons, learning rate, and number of epochs
hidden_neurons = 4
learning_rate = 0.1
epochs = 10000

# Apply the delta learning rule to solve the XOR problem with a two-layered network
synaptic_weights1, synaptic_weights2 = delta_learning_rule_two_layer(input_data, target_output, hidden_neurons, learning_rate, epochs)

# Print the trained weights
print("Trained synaptic weights for the first layer:")
print(synaptic_weights1)
print("Trained synaptic weights for the second layer:")
print(synaptic_weights2)



Trained synaptic weights for the first layer:
[[ 1.01820484  2.54064388 -5.17301449 -5.47102891]
 [ 0.19820196 -5.75544506 -5.23444398  2.3442128 ]]
Trained synaptic weights for the second layer:
[[ -3.42859423]
 [  5.2697904 ]
 [-11.05641271]
 [  4.72594872]]


   - XOR problemi için giriş ve hedef çıktı verileri tanımlanır.
   - Gizli katman nöron sayısı, öğrenme oranı ve epoch sayısı belirlenir.
   - `delta_learning_rule_two_layer` fonksiyonu kullanılarak ağ eğitilir ve eğitilmiş ağırlıklar elde edilir.

Ağırlık matrisleri, katmanlar arasındaki bağlantıları ve nöronlar arasındaki ağırlıkları temsil eder:
- **İlk Katman (Girdi Katmanı - Gizli Katman)**:
  - Ağırlık matrisi `synaptic_weights1`, 2x4 boyutundadır. Bu, 2 giriş nöronundan 4 gizli nörona bağlantıları ifade eder.
  ```python
  synaptic_weights1 = 2 * np.random.random((2, hidden_neurons)) - 1
  ```
  Çıktı:
  ```
  [[ 1.01820484  2.54064388 -5.17301449 -5.47102891]
   [ 0.19820196 -5.75544506 -5.23444398  2.3442128 ]]
  ```

- **İkinci Katman (Gizli Katman - Çıktı Katmanı)**:
  - Ağırlık matrisi `synaptic_weights2`, 4x1 boyutundadır. Bu, 4 gizli nörondan 1 çıktı nöronuna bağlantıları ifade eder.
  ```python
  synaptic_weights2 = 2 * np.random.random((hidden_neurons, 1)) - 1
  ```
  Çıktı:
  ```
  [[ -3.42859423]
   [  5.2697904 ]
   [-11.05641271]
   [  4.72594872]]
  ```

### Neden Tek Dizi?
Çıktı katmanı yalnızca tek bir nörona sahip olduğundan, bu nöronun tüm gizli katmandaki nöronlarla olan bağlantı ağırlıklarını içeren `synaptic_weights2` matrisi, `hidden_neurons` sayısı kadar satıra ve 1 sütuna sahiptir. Bu nedenle, tek bir sütundan oluşan bir dizi gibi görünür.


In [ ]:
# Test the neural network with new inputs
new_inputs = np.array([[0.9, 0.2],
                       [0.1, 0.2],
                       [0.8, 0.5],
                       [0.5, 0.5]])

hidden_layer_output = sigmoid(np.dot(new_inputs, synaptic_weights1))
predicted_output = sigmoid(np.dot(hidden_layer_output, synaptic_weights2))

print("Predicted output for new inputs:")
print(predicted_output)

Predicted output for new inputs:
[[0.82191718]
 [0.51209744]
 [0.33347586]
 [0.35837418]]


- Yeni giriş verileri tanımlanır.
- Eğitim sırasında öğrenilen ağırlıklar kullanılarak bu yeni girişler için ileri yayılım yapılır ve tahmini çıktılar hesaplanır.
- Yeni girişler için tahmin edilen çıktılar yazdırılır.


Bu değerler, sigmoid aktivasyon fonksiyonundan elde edildiği için [0, 1] aralığında bulunur ve bu da olasılık anlamına gelir. XOR probleminin çıktılarını tahmin ederken, çıktılar 0 veya 1 olmalıdır. Bu durumda:

1. **Yeni Girdi: [0.9, 0.2]**
   - Tahmin Edilen Çıktı: 0.82191718
   - Bu değer 0'a daha yakındır ve sinir ağı bu girişler için çıktının 1 olma olasılığının yüksek olduğunu tahmin etmektedir.
   
2. **Yeni Girdi: [0.1, 0.2]**
   - Tahmin Edilen Çıktı: 0.51209744
   - Bu değer 0.5'e yakındır ve sinir ağı bu girişler için çıktının 0 veya 1 olabileceğine dair belirsizlik içindedir.
   
3. **Yeni Girdi: [0.8, 0.5]**
   - Tahmin Edilen Çıktı: 0.33347586
   - Bu değer 0'a daha yakındır ve sinir ağı bu girişler için çıktının 0 olma olasılığının yüksek olduğunu tahmin etmektedir.

4. **Yeni Girdi: [0.5, 0.5]**
   - Tahmin Edilen Çıktı: 0.35837418
   - Bu değer de 0'a daha yakındır ve sinir ağı bu girişler için çıktının 0 olma olasılığının yüksek olduğunu tahmin etmektedir.

### Yorumlama
- **0.5'ten büyük değerler**, ağın bu girdi için 1 çıkmasını beklediğini gösterir.
- **0.5'ten küçük değerler**, ağın bu girdi için 0 çıkmasını beklediğini gösterir.

XOR problemi için:
- Girdi: (0, 0) -> Çıktı: 0
- Girdi: (0, 1) -> Çıktı: 1
- Girdi: (1, 0) -> Çıktı: 1
- Girdi: (1, 1) -> Çıktı: 0

Ancak, tahmin edilen değerler arasında bir belirsizlik veya hata olabilir çünkü yeni girişler eğitim verilerinden farklıdır. Eğitim verileriyle tamamen örtüşmeyen yeni girişler için ağın çıktıları daha belirsiz olabilir. Bu yüzden bazı tahminler 0.5'e yakın olabilir.
